py/server/deephaven/experimental/partitioned_table_service.py

```diff
@@ -0,0 +1,326 @@
+_JTableLocationKeyImpl = jpy.get_type("io.deephaven.extensions.barrage.util.PythonTableDataService$TableLocationKeyImpl")
+
+
+class TableKey:
+    """A key that identifies a table. The key should be unique for each table. The key can be any Python object and
+    should include sufficient information to uniquely identify the table for the backend service."""
+
+    def __init__(self, key: Any):
+        self._key = key
+
+    @property
+    def key(self) -> Any:
+        """The user defined key that identifies the table."""
+        return self._key
+
+
+class PartitionedTableLocationKey:
+    """A key that identifies a specific partition of a table. The key should be unique for each partition of the table.
+    The key can be any Python object and should include sufficient information to uniquely identify the partition for
+    the backend service to fetch the partition data.
+    """
+
+    def __init__(self, pt_location_key: Any):
+        self._pt_location_key = pt_location_key
+
+    @property
+    def pt_location_key(self) -> Any:
+        return self._pt_location_key
+
+
+
+class PartitionedTableServiceBackend(ABC):
+    """An interface for a backend service that provides access to partitioned data."""
+
+    @abstractmethod
+    def table_schema(self, table_key: TableKey) -> Tuple[pa.Schema, Optional[pa.Schema]]:
+        """ Returns the table schema and optionally the schema for the partition columns for the table with the given
+        table key.
+        The table schema is not required to include the partition columns defined in the partition schema. The
+        partition columns are limited to primitive types and strings.
+
+        Args:
+            table_key (TableKey): the table key
+
+        Returns:
+            Tuple[pa.Schema, Optional[pa.Schema]]: a tuple of the table schema and the optional schema for the partition
+                columns
+        """
+        pass
+
+    @abstractmethod
+    def existing_partitions(self, table_key: TableKey,
+                            callback: Callable[[PartitionedTableLocationKey, Optional[pa.Table]], None]) -> None:
+        """ Provides a callback for the backend service to pass the existing partitions for the table with the given
+        table key. The 2nd argument of the callback is an optional pa.Table that contains the values for the partitions.
+        The schema of the table should match the optional partition schema returned by table_schema() for the table_key.
+        The table should have a single row for the particular partition location key provided in the 1st argument,
+        with the values for the partition columns in the row.
+
+        This is called for tables created when ：meth:`PythonTableDataService.make_table` is called with live=False
+
+        Args:
+            table_key (TableKey): the table key
+            callback (Callable[[PartitionedTableLocationKey, Optional[pa.Table]], None]): the callback function
+        """
+        pass
```

Comment:
raise `NotImplementedError`?  Same for other cases

Reply:
No need. Python will complain about an incomplete subclass implementation on instantiation.